caffe2/operators/conv_dgrad_op_cudnn.cc

```diff
@@ -0,0 +1,291 @@
+    // Set the output
+    SetTensorNdDescriptorWithGroup<T_DX>(
+        X.ndim(), top_desc_, in.N, M, out.H, out.W, out.D);
+
+    // Set the convolution descriptor
+    SetConvolutionDescriptor<MATH>(
+          kernel_.size(),
+          conv_desc_);
+
+    // Set the workspace
+    size_t ws_size;
+
+    // Pick dX algo if needed
+    if (force_algo_[ALGO_DGRAD] >= 0) {
+      algo_ = (cudnnConvolutionBwdDataAlgo_t)force_algo_[ALGO_DGRAD];
+    } else if (deterministic_) {
+      algo_ = CUDNN_CONVOLUTION_BWD_DATA_ALGO_1;
+    } else if (exhaustive_search_) {
+      algo_ =
+          algo_cache_.getAlgorithm(X.dims(), filter.dims(), [&]() {
+            VLOG(1) << "CUDNN Convolution bwd: doing data exhaustive search.";
+            int returned_algo_count;
+
+            std::array<
+                cudnnConvolutionBwdDataAlgoPerf_t,
+                kNUM_CUDNN_BWD_DATA_ALGS>
+                data_perf_stat;
+            cudnn_wrapper_.with_cudnn_state(
+                cudnn_state_, [&](CuDNNState* state) {
+                  auto* dX =
+                      Output(INPUT_GRAD);
+                  dX->ResizeLike(X);
+                  const T_W* filter_data = filter.template data<T_W>();
+                  const T_DY* dYdata = dY.template data<T_DY>();
+                  T_DX* dXdata = dX->template mutable_data<T_DX>();
+                  CUDNN_ENFORCE(cudnnFindConvolutionBackwardDataAlgorithmEx(
+                      state->cudnn_handle(),
+                      filter_desc_,
+                      filter_data,
+                      top_desc_,
+                      dYdata,
+                      conv_desc_,
+                      bottom_desc_,
+                      dXdata,
+                      kNUM_CUDNN_BWD_DATA_ALGS,
+                      &returned_algo_count,
+                      data_perf_stat.data(),
+                      state->workspace().get(cudnn_ws_nbytes_limit_),
+                      cudnn_ws_nbytes_limit_));
+                });
+
+            LogCuDNNPerfStats(data_perf_stat, returned_algo_count);
+            return data_perf_stat[0].algo;
+          });
+    } else {
+      CUDNN_ENFORCE(cudnnGetConvolutionBackwardDataAlgorithm(
+          cudnn_wrapper_.inline_cudnn_handle(),
+          filter_desc_,
+          top_desc_,
+          conv_desc_,
+          bottom_desc_,
+          CUDNN_CONVOLUTION_BWD_DATA_SPECIFY_WORKSPACE_LIMIT,
+          cudnn_ws_nbytes_limit_,
+          &algo_));
+    }
+
+    // get workspace for backwards data algorithm
+    CUDNN_ENFORCE(cudnnGetConvolutionBackwardDataWorkspaceSize(
+        cudnn_wrapper_.inline_cudnn_handle(),
+        filter_desc_,
+        top_desc_,
+        conv_desc_,
+        bottom_desc_,
+        algo_,
+        &ws_size));
+    cudnn_ws_nbytes_ = ws_size;
+
+    VLOG(1) << "CuDNN bwd data algorithm: " << algo_;
+    VLOG(1) << "CuDNN workspace size: " << cudnn_ws_nbytes_;
+  }
+
+  std::stringstream cudnnParamStr;
+  if (VLOG_IS_ON(2)) {
+    cudnnParamStr
```

Comment:
I think that need to be member function 
`string CudnnConvOpBase::GetcuDNNParamStr(DimensionParam)`
So you can reuse it below in the forward pass.